File: Packages/LineScopeCore/Sources/LineScopeCore/Spectrum.swift
import Accelerate
import Foundation

public enum WindowFunction: String, CaseIterable, Identifiable, Sendable {
    case none
    case hann

    public var id: String { rawValue }
    public var displayName: String { self == .none ? "Rectangular" : "Hann" }
}

public struct SpectrumResult: Sendable {
    /// Bin frequencies in cycles per pixel of the sampled image (Nyquist ≈ 0.5).
    public var cyclesPerPixel: [Double]
    /// Bin frequencies in cycles over the whole line; comparable across resampled copies.
    public var cyclesPerLine: [Double]
    /// Single-sided amplitude: a sine of amplitude A peaks at ≈ A.
    public var amplitudes: [Double]
    public var fftSize: Int
    public var sampleCount: Int

    public static let empty = SpectrumResult(cyclesPerPixel: [], cyclesPerLine: [], amplitudes: [], fftSize: 0, sampleCount: 0)

    /// Index of the largest non-DC bin.
    public var peakIndex: Int? {
        guard amplitudes.count > 1 else { return nil }
        return amplitudes.indices.dropFirst().max { amplitudes[$0] < amplitudes[$1] }
    }
}

public enum SpectrumAnalyzer {
    /// Real FFT of a profile, zero-padded to the next power of two.
    /// - Parameter spacing: distance between samples, in pixels.
    public static func compute(
        _ values: [Float], spacing: Double, window: WindowFunction, removeMean: Bool
    ) -> SpectrumResult {
        let n = values.count
        guard n >= 2 else { return .empty }
        let log2n = vDSP_Length(Int(ceil(log2(Double(n)))))
        let fftSize = 1 << Int(log2n)
        let half = fftSize / 2

        var signal = values
        if removeMean {
            let mean = vDSP.mean(signal)
            signal = vDSP.add(-mean, signal)
        }
        var windowSum = Float(n)
        if window == .hann {
            // Symmetric Hann over the n real samples.
            var w = [Float](repeating: 0, count: n)
            for i in 0..<n {
                w[i] = 0.5 - 0.5 * cos(2 * Float.pi * Float(i) / Float(n - 1))
            }
            signal = vDSP.multiply(signal, w)
            windowSum = vDSP.sum(w)
        }
        var padded = [Float](repeating: 0, count: fftSize)
        padded.replaceSubrange(0..<n, with: signal)

        var real = [Float](repeating: 0, count: half)
        var imag = [Float](repeating: 0, count: half)
        guard let setup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2)) else { return .empty }
        defer { vDSP_destroy_fftsetup(setup) }

        real.withUnsafeMutableBufferPointer { rp in
            imag.withUnsafeMutableBufferPointer { ip in
                var split = DSPSplitComplex(realp: rp.baseAddress!, imagp: ip.baseAddress!)
                padded.withUnsafeBytes { raw in
                    vDSP_ctoz(raw.bindMemory(to: DSPComplex.self).baseAddress!, 2, &split, 1, vDSP_Length(half))
                }
                vDSP_fft_zrip(setup, &split, 1, log2n, FFTDirection(FFT_FORWARD))
            }
        }

        // vDSP's packed output is scaled by 2; DC is in real[0] and Nyquist in imag[0].
        var amplitudes = [Double](repeating: 0, count: half + 1)
        let norm = Double(max(windowSum, 1e-12))
        amplitudes[0] = abs(Double(real[0]) / 2) / norm
        amplitudes[half] = abs(Double(imag[0]) / 2) / norm
        for k in 1..<max(half, 1) {
            let re = Double(real[k]) / 2
            let im = Double(imag[k]) / 2
            amplitudes[k] = 2 * (re * re + im * im).squareRoot() / norm
        }

        let lineLength = spacing * Double(n - 1)
        var cpp = [Double](repeating: 0, count: half + 1)
        var cpl = [Double](repeating: 0, count: half + 1)
        for k in 0...half {
            cpp[k] = Double(k) / (Double(fftSize) * spacing)
            cpl[k] = cpp[k] * lineLength
        }
        return SpectrumResult(
            cyclesPerPixel: cpp, cyclesPerLine: cpl, amplitudes: amplitudes,
            fftSize: fftSize, sampleCount: n
        )
    }
}

File: Windows/LineScope.Core/Spectrum.cs
using System.Numerics;

namespace LineScope.Core;

public enum WindowFunction { None, Hann }

public static class WindowFunctionInfo
{
    public static readonly WindowFunction[] All = [WindowFunction.None, WindowFunction.Hann];

    public static string DisplayName(this WindowFunction w) => w == WindowFunction.None ? "Rectangular" : "Hann";
}

/// <param name="CyclesPerPixel">Bin frequencies in cycles per pixel of the sampled image (Nyquist ≈ 0.5).</param>
/// <param name="CyclesPerLine">Bin frequencies in cycles over the whole line; comparable across resampled copies.</param>
/// <param name="Amplitudes">Single-sided amplitude: a sine of amplitude A peaks at ≈ A.</param>
public sealed record SpectrumResult(
    double[] CyclesPerPixel, double[] CyclesPerLine, double[] Amplitudes, int FftSize, int SampleCount)
{
    public static readonly SpectrumResult Empty = new([], [], [], 0, 0);

    /// <summary>Index of the largest non-DC bin.</summary>
    public int? PeakIndex
    {
        get
        {
            if (Amplitudes.Length <= 1) return null;
            int best = 1;
            for (int i = 2; i < Amplitudes.Length; i++)
                if (Amplitudes[best] < Amplitudes[i]) best = i;
            return best;
        }
    }
}

public static class SpectrumAnalyzer
{
    /// <summary>Real FFT of a profile, zero-padded to the next power of two.</summary>
    /// <param name="spacing">Distance between samples, in pixels.</param>
    public static SpectrumResult Compute(float[] values, double spacing, WindowFunction window, bool removeMean)
    {
        int n = values.Length;
        if (n < 2) return SpectrumResult.Empty;
        int log2n = (int)Math.Ceiling(Math.Log2(n));
        int fftSize = 1 << log2n;
        int half = fftSize / 2;

        var signal = (float[])values.Clone();
        if (removeMean)
        {
            float mean = signal.Sum() / n;
            for (int i = 0; i < n; i++) signal[i] -= mean;
        }
        float windowSum = n;
        if (window == WindowFunction.Hann)
        {
            // Symmetric Hann over the n real samples.
            windowSum = 0;
            for (int i = 0; i < n; i++)
            {
                float w = 0.5f - 0.5f * MathF.Cos(2 * MathF.PI * i / (n - 1));
                signal[i] *= w;
                windowSum += w;
            }
        }

        var buffer = new Complex[fftSize];
        for (int i = 0; i < n; i++) buffer[i] = signal[i];
        Fft.Forward(buffer);

        var amplitudes = new double[half + 1];
        double norm = Math.Max(windowSum, 1e-12);
        amplitudes[0] = buffer[0].Magnitude / norm;
        amplitudes[half] = buffer[half].Magnitude / norm;
        for (int k = 1; k < Math.Max(half, 1); k++)
            amplitudes[k] = 2 * buffer[k].Magnitude / norm;

        double lineLength = spacing * (n - 1);
        var cpp = new double[half + 1];
        var cpl = new double[half + 1];
        for (int k = 0; k <= half; k++)
        {
            cpp[k] = k / (fftSize * spacing);
            cpl[k] = cpp[k] * lineLength;
        }
        return new SpectrumResult(cpp, cpl, amplitudes, fftSize, n);
    }
}

/// <summary>In-place iterative radix-2 Cooley-Tukey FFT (replaces vDSP on macOS).</summary>
internal static class Fft
{
    /// <summary>Forward DFT, X[k] = Σ x[j]·e^(−2πi·jk/N), unscaled. The length must be a power of two.</summary>
    public static void Forward(Complex[] a)
    {
        int n = a.Length;
        if (n <= 1) return;
        if ((n & (n - 1)) != 0) throw new ArgumentException("FFT length must be a power of two.");

        // Bit-reversal permutation.
        for (int i = 1, j = 0; i < n; i++)
        {
            int bit = n >> 1;
            for (; (j & bit) != 0; bit >>= 1) j ^= bit;
            j ^= bit;
            if (i < j) (a[i], a[j]) = (a[j], a[i]);
        }

        for (int len = 2; len <= n; len <<= 1)
        {
            double angle = -2 * Math.PI / len;
            var wLen = new Complex(Math.Cos(angle), Math.Sin(angle));
            int halfLen = len >> 1;
            for (int i = 0; i < n; i += len)
            {
                Complex w = Complex.One;
                for (int k = 0; k < halfLen; k++)
                {
                    Complex u = a[i + k];
                    Complex v = a[i + k + halfLen] * w;
                    a[i + k] = u + v;
                    a[i + k + halfLen] = u - v;
                    w *= wLen;
                }
            }
        }
    }
}
